packages/astro/src/core/errors/errors-data.ts

```diff
@@ -628,6 +628,94 @@ See https://docs.astro.build/en/guides/server-side-rendering/ for more informati
 		code: 3030,
 		message: 'The response has already been sent to the browser and cannot be altered.',
 	},
+
+	/**
+	 * @docs
+	 * @description
+	 * Thrown when the middleware does not return any data or call the `next` function.
+	 *
+	 * For example:
+	 * ```ts
+	 * import {defineMiddleware} from "astro/middleware";
+	 * export const onRequest = defineMiddleware((context, _) => {
+	 * 	// doesn't return anything or call `next`
+	 * 	context.locals.someData = false;
+	 * });
+	 * ```
+	 */
+	MiddlewareNoDataOrNextCalled: {
+		title: "The middleware didn't return a response or call `next`",
+		code: 3031,
+		message: 'The middleware needs to either return a `Response` object or call the `next` function.',
+	},
+
+	/**
+	 * @docs
+	 * @description
+	 * Thrown in development mode when middleware returns something that is not a `Response` object.
+	 *
+	 * For example:
+	 * ```ts
+	 * import {defineMiddleware} from "astro/middleware";
+	 * export const onRequest = defineMiddleware(() => {
+	 *   return "string"
+	 * });
+	 * ```
+	 */
+	MiddlewareNotAResponse: {
+		title: 'The middleware returned something that is not a `Response` object',
+		code: 3032,
+		message:
+			'Any data returned from middleware must be a valid `Response` object.',
+	},
+
+	/**
+	 * @docs
+	 * @description
+	 *
+	 * Thrown in development mode, when `locals` are overridden with something that is not an object
+	 *
+	 * For example:
+	 * ```ts
+	 * import {defineMiddleware} from "astro/middleware";
+	 * export const onRequest = defineMiddleware((context, next) => {
+	 *   context.locals = 1541;
+	 *   return next();
+	 * });
+	 * ```
+	 */
+	LocalsNotAnObject: {
+		title: 'Value assigned to `locals` is not accepted.',
+		code: 3033,
+		message: `The \`locals\` can only be assigned to an object. Other values like numbers, strings, etc. are not accepted.`,
+		hint: 'If you tried to remove some information from the `locals` object, try to use `delete` or set the property to `undefined`.',
+	},
+
+	/**
+	 * @docs
+	 * @description
+	 * Thrown in development mode, when a user attempts to store in `local` something that is not serializable.
+	 *
+	 * For example:
+	 * ```ts
+	 * import {defineMiddleware} from "astro/middleware";
+	 * export const onRequest = defineMiddleware((context, next) => {
+	 *   context.locals = {
+	 *     foo() {
+	 *       alert("Hello world!")
+	 *     }
+	 *   };
+	 *   return next();
+	 * });
+	 * ```
+	 */
+	LocalsNotSerializable: {
+		title: '`Astro.locals` are not serializable.',
+		code: 3034,
+		message: (href: string) => {
+			return `The information stored in \`Astro.locals\` are not serializable when visiting "${href}" path.\nMake sure you store only data that are serializable.`;
```

Comment:
```suggestion
			return `The information stored in \`Astro.locals\` is not serializable when visiting "${href}" path.\nMake sure you store only serializable data.`;
```
Information is singular, and I'm not sure about "when visiting... path" If that renders to e.g. "when visiting /about/ path" then that sentence wouldn't read very well in English.

What exactly is the signifiance of including the part about the path? Would the warning work without it, or is the message depending on knowing which path is causing the error? In that case, instead of "when visiting the path" what is the actual problem? The route can't be created? The path doesn't exist? 

> The information stored in Astro.locals for the path "${href} is not serializable. Mke sure...

Would something closer to that be possible?

Reply:
> What exactly is the signifiance of including the part about the path?

The logic of the middleware can change based on the page the user is visiting. So having a path can help the developer understand the issue.

> > The information stored in Astro.locals for the path "${href} is not serializable. Make sure...
> 
> Would something closer to that be possible?

That's perfect!